### ISSUE TITLE
Possible dropped messages during Publish

### ISSUE THREAD

**Issue by danneaves-ee** [olivertappin/scoop#4]
https://github.com/olivertappin/scoop/blob/57f180720ea31cf21860aa751a428d2b7d597d1c/scoop/scoop.go#L178

https://godoc.org/github.com/streadway/amqp#Channel.Publish it states;

```Since publishings are asynchronous, any undeliverable message will get returned by the server. Add a listener with Channel.NotifyReturn to handle any undeliverable message when calling publish with either the mandatory or immediate parameters as true.```

and...

```This can return an error when the channel, connection or socket is closed. The error or lack of an error does not indicate whether the server has received this publishing.```

This leads to the potential for dropped messages.

**Comment by olivertappin** [comment]
Thanks @danneaves-ee. This is a good callout and one worth discussing.

I think there are two concerns to address here:

1. Potential for dropped messages during a clean use of Scoop (i.e. no fatals)
2. Potential for dropped messages after a fatal

The first is most concerning, as users may not be aware of potential dropped messages. This is high priority and should be addressed immediately.

The second is less concerning as the first, because the user of Scoop will be aware of the error, and could have bigger problems if a connection is dropped. That doesn't mean to say it isn't important, it's just that it's visually obvious to the user at the point of the fatal error.

Looking at this comment from the [example](https://godoc.org/github.com/streadway/amqp#Channel.Publish) within the documentation:
> Since publish is asynchronous this can happen if the network connection is reset or if the server has run out of resources.
https://godoc.org/github.com/streadway/amqp#Channel.Publish

It suggests that we only reach that if the publish itself errors. Then, looking at the code within Scoop, I can see we don't actually acknowledge the message until it's been successfully published.

The code below is taken from the example provided:

```
// This is not a mandatory delivery, so it will be dropped if there are no
// queues bound to the logs exchange.
err = c.Publish("logs", "info", false, false, msg)
if err != nil {
    // Since publish is asynchronous this can happen if the network connection
    // is reset or if the server has run out of resources.
    log.Fatalf("basic.publish: %v", err)
}
```

Which seems similar to what we have in Scoop today:

```
err = publisherChannel.Publish(
    *exchange,    // exchange
    toQueue.Name, // routing key
    true,         // mandatory
    false,        // immediate
    amqp.Publishing{
        ContentType: "text/plain",
        Body:        []byte(d.Body),
    })

failOnError(err, "Failed to deliver message")
```

Meaning, an error message would show within the command prompt (`Failed to deliver message`), if the `err != nil` scenario were to happen.

Through the normal usage of Scoop alone, without any fatal errors, do you think it is susceptible to drop messages? Even if it had thrown a fatal error and the consumer stopped, because we wouldn't have acknowledged the message, it would have appeared back in the initial queue. Do you believe this not to be the case?

For an initial set-up, ignoring dropped connections during the running of a consumer (just for a moment), when we declare the queue, the queue will not be created successfully if all the parameters match. For example, connecting to a queue which has a different routing key to the one configured by the command. Which is essentially human error during the command creation.

Attempting to address the other point, the manual goes on to talk about `mandatory` and `exclusive` flags, where the use of `Channel.NotifyReturn` is suggested:

> Since publishings are asynchronous, any undeliverable message will get returned by the server. Add a listener with Channel.NotifyReturn to handle any undeliverable message when calling publish with either the mandatory or immediate parameters as true.
https://godoc.org/github.com/streadway/amqp#Channel.Publish

What's interesting is the following section after that:
> It is possible for publishing to not reach the broker if the underlying socket is shut down without pending publishing packets being flushed from the kernel buffers. The easy way of making it probable that all publishings reach the server is to always call `Connection.Close` before terminating your publishing application. The way to ensure that all publishings reach the server is to add a listener to `Channel.NotifyPublish` and put the channel in confirm mode with `Channel.Confirm`. Publishing delivery tags and their corresponding confirmations start at 1. Exit when all publishings are confirmed.
https://godoc.org/github.com/streadway/amqp#Channel.Publish

The first part suggests that it's still not full-proof, as it's only probable.
> The easy way of making it probable that all publishings reach the server

However, I think we should still make the necessary changes to close the connection before terminating the Scoop consumer.

And the second talks about a different approach, which doesn't require the use of `mandatory` and `exclusive` flags:
> The way to ensure that all publishings reach the server is to add a listener to `Channel.NotifyPublish` and put the channel in confirm mode with `Channel.Confirm`.

So, what is it that you are suggesting? I feel like there are two options:

1. Add a listener with `Channel.NotifyPublish` and put the channel in confirm mode with `Channel.Confirm`
2. Add a listener with `Channel.NotifyReturn` to handle any undeliverable message when calling publish with either the `mandatory` or `immediate` parameters as `true`.

Or, do we need to carry out both? 🤔

**Comment by danneaves-ee** [comment]
This is clearly a complex subject, and not being familiar specifically with `go` it's hard to comment on the specific implementation. However, if I was to provide a suggestion, I would be looking to follow the example shown here; https://godoc.org/github.com/streadway/amqp#example-Channel-Confirm-Bridge

The example seems to implement your option 1, where the source is only `ack|nack`ed on the result of the `Confirmation`.

In addition - I would think your listener for `Channel.NotifyReturn` should always `Nack` the source.

So, in summary, I think we need both 1 & 2 for ultimate integrity.

**Comment by olivertappin** [comment]
This has been addressed in [2.1.0](https://github.com/olivertappin/scoop/releases/tag/2.1.0). Additional reliability has been introduced as discussed. For more information, see [these changes](https://github.com/olivertappin/scoop/commit/0c30c8191080ee881c8e48fc21aeebf069cc86a5).
